azure-iot-device/tests/iothub/test_sync_handler_manager.py

```diff
@@ -359,20 +463,179 @@ def handler(arg):
+        # Set the handler back to None
+        setattr(handler_manager, handler_name, None)
+        # Thread is still maintained on the manager and alive
+        assert handler_manager._client_event_runner is t
+        assert t.is_alive()
+
+    @pytest.mark.it(
+        "Does not delete, remove, or replace the Thread for the Client Event handler runner when updated with a new function value"
+    )
+    def test_handler_update(self, handler_name, handler_manager, handler):
+        # Set handler
+        setattr(handler_manager, handler_name, handler)
+        # Thread has been created and is alive
+        t = handler_manager._client_event_runner
+        assert isinstance(t, threading.Thread)
+        assert t.is_alive()
+
+        # Set new handler
+        def new_handler(arg):
+            pass
+
+        setattr(handler_manager, handler_name, new_handler)
+
+        # Thread is still maintained on the manager and alive
+        assert handler_manager._client_event_runner is t
+        assert t.is_alive()
+
+    @pytest.mark.it(
+        "Is invoked by the runner only when the Client Event Inbox receives a matching Client Event, passing any arguments to the handler"
+    )
+    def test_handler_invoked(self, mocker, handler_name, handler_manager, inbox, event):
+        # Set the handler
+        mock_handler = mocker.MagicMock()
+        setattr(handler_manager, handler_name, mock_handler)
+        # Handler has not been called
+        assert mock_handler.call_count == 0
+
+        # Add the event to the client event inbox
+        inbox.put(event)
+        time.sleep(0.1)
+
+        # Handler has been called with the arguments from the event
+        assert mock_handler.call_count == 1
+        assert mock_handler.call_args == mocker.call(*event.args_for_user)
+
+        # Add non-matching event to the client event inbox
+        non_matching_event = client_event.ClientEvent("NON_MATCHING_EVENT")
+        inbox.put(non_matching_event)
+        time.sleep(0.1)
+
+        # Handler has not been called again
+        assert mock_handler.call_count == 1
+
+    @pytest.mark.it(
+        "Is invoked by the runner every time the Client Event Inbox receives a matching Client Event"
+    )
+    def test_handler_invoked_multiple(self, handler_name, handler_manager, inbox, event):
+        # Set the handler
+        mock_handler = ThreadsafeMock()
+        setattr(handler_manager, handler_name, mock_handler)
+        # Handler has not been called
+        assert mock_handler.call_count == 0
+
+        # Add 5 matching events to the corresponding inbox, triggering the handler
+        for _ in range(5):
+            inbox.put(event)
+        time.sleep(0.2)
+
+        # Handler has been called 5 times
+        assert mock_handler.call_count == 5
+
+    @pytest.mark.it(
+        "Sends a HandlerManagerException to the background exception handler if any exception is raised during its invocation"
+    )
+    def test_exception_in_handler(
```

Comment:
This functionality will of course have to change once the background exception handler is actually implemented